File: youtube/lat7-diabetes/try2.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns 
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_selection import RFE
import statsmodels.api as sm 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor


diabet = pd.read_csv('ML/youtube/lat7-diabetes/diabetes.csv')
print(diabet)
print(diabet.info())
print(diabet.describe())
print(diabet.isnull().sum())

for col in diabet:
    print(col)
    print(diabet[col].unique())
    print('\n\n')


fig, ax = plt.subplots(2, 5, figsize=(15, 10))
sns.scatterplot(x=diabet.index, y=diabet['Pregnancies'], ax=ax[0, 0], color='blue')
sns.scatterplot(x=diabet.index, y=diabet['Glucose'], ax=ax[0, 1], color='green')
sns.scatterplot(x=diabet.index, y=diabet['BloodPressure'], ax=ax[0, 2], color='orange')
sns.scatterplot(x=diabet.index, y=diabet['SkinThickness'], ax=ax[0, 3], color='red')
sns.scatterplot(x=diabet.index, y=diabet['Insulin'], ax=ax[1, 0], color='blue')
sns.scatterplot(x=diabet.index, y=diabet['BMI'], ax=ax[1, 1], color='green')
sns.scatterplot(x=diabet.index, y=diabet['DiabetesPedigreeFunction'], ax=ax[1, 2], color='orange')
sns.scatterplot(x=diabet.index, y=diabet['Age'], ax=ax[1, 3], color='red')
sns.scatterplot(x=diabet.index, y=diabet['Outcome'], ax=ax[1, 4], color='red')
plt.tight_layout()
plt.show()

q1 = diabet['Pregnancies'].quantile(0.25)
q3 = diabet['Pregnancies'].quantile(0.75)
iqr = q3 - q1
diabet = diabet[(diabet['Pregnancies'] >= q1 - 1.5*iqr) & (diabet['Pregnancies'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['Glucose'], 25)
q3 = np.percentile(diabet['Glucose'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['Glucose'] >= q1 - 1.5*iqr) & (diabet['Glucose'] <= q3 + 1.5*iqr )]

q1 = diabet['BloodPressure'].quantile(0.25)
q3 = diabet['BloodPressure'].quantile(0.75)
iqr = q3 - q1
diabet = diabet[(diabet['BloodPressure'] >= q1 - 1.5*iqr) & (diabet['BloodPressure'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['SkinThickness'], 25)
q3 = np.percentile(diabet['SkinThickness'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['SkinThickness'] >= q1 - 1.5*iqr) & (diabet['SkinThickness'] <= q3 + 1.5*iqr )]

q1 = diabet['Insulin'].quantile(0.25)
q3 = diabet['Insulin'].quantile(0.75)
iqr = q3 - q1
diabet = diabet[(diabet['Insulin'] >= q1 - 1.5*iqr) & (diabet['Insulin'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['BMI'], 25)
q3 = np.percentile(diabet['BMI'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['BMI'] >= q1 - 1.5*iqr) & (diabet['BMI'] <= q3 + 1.5*iqr )]

q1 = diabet['DiabetesPedigreeFunction'].quantile(0.25)
q3 = diabet['DiabetesPedigreeFunction'].quantile(0.75)
iqr = q3 - q1
diabet = diabet[(diabet['DiabetesPedigreeFunction'] >= q1 - 1.5*iqr) & (diabet['DiabetesPedigreeFunction'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['Age'], 25)
q3 = np.percentile(diabet['Age'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['Age'] >= q1 - 1.5*iqr) & (diabet['Age'] <= q3 + 1.5*iqr )]

q1 = np.percentile(diabet['Outcome'], 25)
q3 = np.percentile(diabet['Outcome'], 75)
iqr = q3 - q1
diabet = diabet[(diabet['Outcome'] >= q1 - 1.5*iqr) & (diabet['Outcome'] <= q3 + 1.5*iqr )]


fig, ax = plt.subplots(2, 5, figsize=(15, 10))
sns.scatterplot(x=diabet.index, y=diabet['Pregnancies'], ax=ax[0, 0], color='blue')
sns.scatterplot(x=diabet.index, y=diabet['Glucose'], ax=ax[0, 1], color='green')
sns.scatterplot(x=diabet.index, y=diabet['BloodPressure'], ax=ax[0, 2], color='orange')
sns.scatterplot(x=diabet.index, y=diabet['SkinThickness'], ax=ax[0, 3], color='red')
sns.scatterplot(x=diabet.index, y=diabet['Insulin'], ax=ax[1, 0], color='blue')
sns.scatterplot(x=diabet.index, y=diabet['BMI'], ax=ax[1, 1], color='green')
sns.scatterplot(x=diabet.index, y=diabet['DiabetesPedigreeFunction'], ax=ax[1, 2], color='orange')
sns.scatterplot(x=diabet.index, y=diabet['Age'], ax=ax[1, 3], color='red')
sns.scatterplot(x=diabet.index, y=diabet['Outcome'], ax=ax[1, 4], color='red')
plt.tight_layout()
plt.show()


numvars = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age']
scaler = MinMaxScaler()
diabet[numvars] = scaler.fit_transform(diabet[numvars])

print(diabet)

df_train, df_test = train_test_split(diabet, train_size = 0.7, test_size=0.3, random_state=100)
print(df_train.head())
print(df_test.head())

corr = df_train.corr()
print(corr)
# plt.figure(figsize=(16, 10))
# sns.heatmap(corr, annot = True, cmap="YlGnBu")
# plt.show()

x_train = df_train
y_train = df_train.pop('Outcome')

k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelLR  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(modelLR, x_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))
modelLR.fit(x_train, y_train)


# sebenarnya tidak dibutuhkan karena jumlah kolomnya sendiri hanya 4
rfe = RFE(estimator=modelLR, n_features_to_select=5)
rfe = rfe.fit(x_train, y_train)
print(rfe.support_)
print(rfe.ranking_)

hasil = pd.DataFrame({
    'features': x_train.columns,
    'support': rfe.support_,
    'rank': rfe.ranking_,
})
print(hasil)

support = x_train.columns[rfe.support_]
not_support = x_train.columns[~rfe.support_]
print(support)
print(not_support)

x_train_rfe = x_train[support]
# x_train_rfe = sm.add_constant(x_train_rfe)
print(x_train_rfe.head())
print(x_train_rfe.shape)

# k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
# modelLR  = LogisticRegression()
# scoring = 'accuracy'
# score = cross_val_score(modelLR, x_train_rfe, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
# print(score)
# print(round(score.mean(), 2))
# modelLR.fit(x_train_rfe, y_train)

# vif = pd.DataFrame()
# x = x_train_rfe
# print(x.shape)

# vif['features'] = x.columns
# vif['VIF'] = [variance_inflation_factor(x.values, i) for i in range(x.shape[1])]
# vif = vif.sort_values('VIF', ascending=False)
# print(vif)

# vif['VIF'] = round(vif['VIF'], 2)
# vif = vif.sort_values('VIF', ascending=False)
# print(vif)

# y_pred = modelLR.predict(x_train_rfe)

# df = pd.DataFrame({
#     'y_train': y_train,
#     'y_pred': y_pred,
# })
# print(df.head())
# print(df[df['y_train'] != df['y_pred']].head())

# akurasi = accuracy_score(y_train, y_pred)
# print(akurasi)
# cm = confusion_matrix(y_train, y_pred)
# print(cm)
# sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
# plt.xlabel('y_pred')
# plt.ylabel('y_test')
# plt.title('confusion matrix')
# plt.show()



# x_test = df_test
# x_test_rfe = df_test[support]
# y_test = df_test.pop('Outcome')
# y_pred = modelLR.predict(x_test_rfe)

# df = pd.DataFrame({
#     'y_test': y_test,
#     'y_pred': y_pred,
# })
# print(df.head())
# print(df[df['y_test'] != df['y_pred']])

# akurasi = accuracy_score(y_test, y_pred)
# print(akurasi)
# cm = confusion_matrix(y_test, y_pred)
# print(cm)
# sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
# plt.xlabel('y_pred')
# plt.ylabel('y_test')
# plt.title('confusion matrix')
# plt.show()

File: youtube/lat7-diabetes/try2.r
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(RMySQL)
library(corrplot)
library(caTools)
library(dplyr)
library(fastDummies)
library(zoo)
library(pROC)
library(ROCR)
library(leaps)
library(car)
library(tidyr)
library(plyr)
library(lubridate)
library(randomForest)
library(caret)
library(patchwork)
library(nnet)
library(klaR) # untuk Naive Bayes
library(rpart) # untuk Decision Trees


df <- read.csv("ML/youtube/lat7-diabetes/diabetes.csv")
print(head(df))
print(dim(df))
print(str(df))
print(summary(df))
print(colSums(is.na(df)))
print(sapply(df, class))
print(sapply(df, function(x) length(unique(na.omit(x)))))

## TIDAK BISA KARENA ADANYA TABRAKAN DENGAN PACKAGE LAINNTA
# library(dplyr)
# library(tidyverse)
# library(ggplot2)
# null1 <- df %>% summarise(across(everything(), n_distinct))
# print(as.data.frame(null1))



p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Pregnancies), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Glucose), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = BloodPressure), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = SkinThickness), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Insulin), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = BMI), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = DiabetesPedigreeFunction), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Age), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Outcome), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# # Menggabungkan plot menggunakan patchwork
plot <- p1+p2+p3+p4+p5+p6+p7+p8+p9
# plot <- (p1 | p2) / (p3 | p4)
print(plot)


q1 <- quantile(df$Pregnancies, 0.25)
q3 <- quantile(df$Pregnancies, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, Pregnancies >= q1 - 1.5 * iqr & Pregnancies <= q3 + 1.5 * iqr)

q1 <- quantile(df$Glucose, 0.25)
q3 <- quantile(df$Glucose, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, Glucose >= q1 - 1.5 * iqr & Glucose <= q3 + 1.5 * iqr)

q1 <- quantile(df$BloodPressure , 0.25)
q3 <- quantile(df$BloodPressure , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, BloodPressure  >= q1 - 1.5 * iqr & BloodPressure  <= q3 + 1.5 * iqr)

q1 <- quantile(df$SkinThickness , 0.25)
q3 <- quantile(df$SkinThickness , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, SkinThickness  >= q1 - 1.5 * iqr & SkinThickness  <= q3 + 1.5 * iqr)

q1 <- quantile(df$Insulin  , 0.25)
q3 <- quantile(df$Insulin  , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, Insulin   >= q1 - 1.5 * iqr & Insulin   <= q3 + 1.5 * iqr)

q1 <- quantile(df$BMI , 0.25)
q3 <- quantile(df$BMI , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, BMI  >= q1 - 1.5 * iqr & BMI  <= q3 + 1.5 * iqr)

q1 <- quantile(df$DiabetesPedigreeFunction , 0.25)
q3 <- quantile(df$DiabetesPedigreeFunction , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, DiabetesPedigreeFunction  >= q1 - 1.5 * iqr & DiabetesPedigreeFunction  <= q3 + 1.5 * iqr)

q1 <- quantile(df$Age, 0.25)
q3 <- quantile(df$Age, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, Age >= q1 - 1.5 * iqr & Age <= q3 + 1.5 * iqr)

q1 <- quantile(df$Outcome , 0.25)
q3 <- quantile(df$Outcome , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, Outcome  >= q1 - 1.5 * iqr & Outcome  <= q3 + 1.5 * iqr)

p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Pregnancies), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Glucose), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = BloodPressure), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = SkinThickness), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Insulin), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = BMI), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = DiabetesPedigreeFunction), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Age), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = Outcome), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# # Menggabungkan plot menggunakan patchwork
plot <- p1+p2+p3+p4+p5+p6+p7+p8+p9
# plot <- (p1 | p2) / (p3 | p4)
print(plot)


print(head(df))

numvars <- c('Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age')
scaler <- preProcess(df[numvars], method = "range")
df[numvars] <- predict(scaler, df[numvars])
print(head(df))

## Set random seed
set.seed(42)
## Create train and test indices
split <- sample.split(df, SplitRatio = 0.7)

## Create train and test dataframes
# df_train <- df[split,]
# df_test <- df[!split,]
df_train <- subset(df, split == TRUE)
df_test <- subset(df, split == FALSE)

## Print first few rows of train and test dataframes
print(head(df_train))
print(head(df_test))
print(dim(df))
print(dim(df_train))
print(dim(df_test))


corr <- cor(df_train)
print(corr)
corplot <- corrplot(corr, method = 'color')
print(corplot)

x_train <- subset(df_train, select = -Outcome)
y_train <- as.factor(df_train$Outcome)


ctrl <- trainControl(method="cv", number=5)
modelrf <- train(x_train, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')


# Membuat kontrol RFE
# control <- rfeControl(functions = leapsFuncs, method = "cv", number =10)
control <- rfeControl(functions = rfFuncs, method = "cv", number =10)
# Menjalankan RFE
rfe_model <- rfe(x = x_train, y = y_train, sizes = c(3), rfeControl = control)
# Mendapatkan hasil support dan ranking
support <- predictors(rfe_model)
ranking <- varImp(rfe_model, scale = FALSE)

print(support)
print(ranking)

# Membuat dataframe hasil
hasil <- data.frame(
    Features = rownames(ranking),
    Support = rownames(ranking) %in% support,
    Rank = ranking$Overall
)

print(hasil)

support <- hasil$Features[hasil$Support == TRUE]
not_support <- hasil$Features[hasil$Support == FALSE]
print(support)
print(not_support)

x_train_rfe <- x_train[support]
print(head(x_train_rfe))

ctrl <- trainControl(method="cv", number=5)
modelrf <- train(x_train_rfe, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')


# Menghitung VIF
# Gabungkan x_train_rfe dan y_train menjadi satu data frame
data_train <- cbind(x_train_rfe, response = y_train)
# Membuat formula model linier dengan semua fitur dalam x_train_rfe
formula <- as.formula(paste("response ~", paste(names(x_train_rfe), collapse = " + ")))
# Membuat model linier menggunakan formula tersebut
model <- glm(formula, data = data_train, family = binomial)
# Menghitung VIF
vif_values <- vif(model)
print(vif_values)

# Membuat data frame dengan hasil VIF
vif_df <- data.frame(
    Features = names(vif_values),
    VIF = vif_values
)
vif_df <- vif_df[order(-vif_df$VIF), ]

print(vif_df)



y_pred <- predict(modelrf, x_train_rfe)
df_pred <- data.frame(
    y_train = y_train,
    y_pred = y_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_train != y_pred)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred, reference = y_train)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred, reference = y_train)
print(cm)


x_test <- subset(df_test, select = -Outcome)
x_test_rfe <- x_test[support]
y_test <- as.factor(df_test$Outcome)


y_pred <- predict(modelrf, x_test_rfe)
df_pred <- data.frame(
    y_test = y_test,
    y_pred = y_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_test != y_pred)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred, reference = y_test)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred, reference = y_test)
print(cm)


